### PR TITLE
Add tensor.to(options)

### PR DESCRIPTION
**Pull request by ssnl**
Actual usage coming in #11253

### CODE REVIEW

File: aten/src/ATen/native/TensorConversions.cpp [hunk 1]
```diff
@@ -16,6 +17,27 @@ static Tensor to_impl(const Tensor& self, const TensorOptions& options, bool non
...
+           " and options.layout set as ", options.layout());
+
+  const auto & device_opt = options.device_opt();
+  const auto & dtype_opt = options.dtype_opt();
+  if ((!device_opt || self.device() == device_opt.value()) &&
+      (!dtype_opt  || self.dtype()  ==  dtype_opt.value()) && !copy) {
+    return self;
+  }
+
+  return to_impl(self, options, non_blocking);
```
Comment (ezyang) [inline aten/src/ATen/native/TensorConversions.cpp]: Lol, I like how it's internally implemented as options.

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -2100,6 +2100,10 @@
   variants: function, method
   device_guard: False
 
+- func: to(Tensor self, TensorOptions options, bool non_blocking=false, bool copy=false) -> Tensor
```
Comment (colesbury) [inline aten/src/ATen/native/native_functions.yaml]: Does this create another python binding overload for `to()`?

**Comment by SsnL** [comment]
@colesbury python .to is manually bound, so I don’t think so.

**Comment by SsnL** [comment]
sigh. need to work out the implicit conversion logic. now it is ambiguous with `DeviceType` arg as it is both convertible to `Device` and `TensorOption`

**Comment by SsnL** [comment]
JIT doesn't like this. I'm putting this aside. May revisit later.

**Pull request closed without merging**